### PR TITLE
relieve some pain: restart `check-for-changes.sh` regularly

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

The script does not need to run every single second. This is not to say
that there is huge resource usage due to running every second, but every
~~ten~~ two seconds is more than enough. ~~This _should_ act like a plaster when
it comes to #2348 -~~ I KNOW THIS IS **NOT** A FIX. ~~But this way, people
can first of all stay on Debian 11 and DMS 10.4.0, secondly, we may get
some more feedback on this mysterious issue due to people staying in DMS
10.4.0.~~ Restarting the script will provide a new PID and should in theory stop the resource leakage. Not sure whether we want to go through with it though, as this will also disguise the error at hand to some degree.

Independently of #2348, increasing the timeout is a very valid concern I had, and on my deployment, I already raised it.

Fixes nothing, but _should_ increase the time before DMS starts using all the RAM on a system.

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
@polarathene would providing the one failing test with more time help here, or do you think the issue is somewhere else entirely?

**Comment by oblitum** [comment]
This is probably related to PR #2157.

**Comment by casperklein** [comment]
> Fixes nothing, but should increase the time before DMS starts using all the RAM on a system.

I think a better approach would be, to restart the changedetector service once a day or so.
Your solution does only slow down a potential memory leak, but will not prevent it. Restarting the service should however "fix" it.

**Comment by georglauterbach** [comment]
Good idea. Will `supervisorctl changedetector restart` suffice here?

**Comment by casperklein** [comment]
Yes, I think so. But haven't tested it. If the PID of the script changes afterwards, it works.

**Comment by casperklein** [comment]
BTW: While reviewing the check-for-changes script, I stumbled upon https://github.com/docker-mailserver/docker-mailserver/blob/7b21db77cc5c2723374131a6f9f1ff21fc7f1859/target/scripts/helper-functions.sh#L245-L254

@georglauterbach @polarathene Do you have an idea, why there is a subshell used + `exec` statement? IMO it should work without those too. I cannot imagine a reason for that.

**Comment by georglauterbach** [comment]
> BTW: While reviewing the check-for-changes script, I stumbled upon
> 
> https://github.com/docker-mailserver/docker-mailserver/blob/7b21db77cc5c2723374131a6f9f1ff21fc7f1859/target/scripts/helper-functions.sh#L245-L254
> 
> @georglauterbach @polarathene Do you have an idea, why there is a subshell used + `exec` statement? IMO it should work without those too. I cannot imagine a reason for that.

I see no reason why the subshell is there either... maybe the writer wanted to be able to not run the SHA-sum when he/she could not change directories, but this is an awful way of doing it. I think this can be done in a better way :D

File: target/scripts/startup/fixes-stack.sh [hunk 1]
```diff
@@ -54,3 +54,9 @@ function _fix_cleanup_spamassassin
     [[ ! -f /CONTAINER_START ]] && _notify 'err' 'Failed to remove SpamAssassin configuration'
   }
 }
+
+function _fix_restart_changedetector_daily
+{
+  _notify 'task' 'Making sure the changedetector is restarted daily'
+  echo 'supervisorctl restart changedetector' >/etc/cron.daily/restart_changedetector
```
Comment (casperklein) [inline target/scripts/startup/fixes-stack.sh:61]: 1. No shebang line?
2. `restart_changedetector` is not executable?

Comment (georglauterbach) [reply target/scripts/startup/fixes-stack.sh:61]: 1. I forgot
2. I forgot

Thanks for pointing this out 👌🏽

**Comment by oblitum** [comment]
> I think this can be done in a better way :D

TBH, this spot in the code is the strange part that makes me feel it's the culprit. I wish that, if you folks could refactor this to something less astonishing, we could then test whether that actually removes the issue, before having a PR applied that could daily kill the problem.

**Comment by georglauterbach** [comment]
> > I think this can be done in a better way :D
> 
> 
> 
> TBH, this spot in the code is the strange part that makes me feel it's the culprit. I wish that, if you folks could refactor this to something less astonishing, we could then test whether that actually removes the issue, before having a PR applied that could daily kill the problem.

I will see where I can provide something here unless @casperklein beats me to it :D

**Comment by casperklein** [comment]
>  function _fix_restart_changedetector_daily

One last thing: I think it's a good idea, to add a small comment referencing the issue, so it's clear, why this function exists.
Otherwise it LGTM 👍

**Comment by oblitum** [comment]
Just to be clear, I suspect of sub shell part because it does a full clone of shell script state (without due reason afaik), and this state cloning may be increasing each time it happens, accumulating over previous state (although I couldn't yet realize what could be linearly increasing).

**Comment by NorseGaud** [comment]
> Just to be clear, I suspect of sub shell part because it does a full clone of shell script state (without due reason afaik), and this state cloning might be increasing each time it happens, accumulating over previous state (although I couldn't yet realize what could be linearly increasing), while leaving leftovers behind in the parent shell process. This can explain why both RAM and CPU usage increases linearly.

I think that's a really good point.

**Comment by NorseGaud** [comment]
I've seen exec used instead of eval but both for the same purpose: allowing interpolation of dynamic variables to happen before execution. Likely why because of the array expansion/interpolation happening. Though, it should work just fine without exec afaik

**Comment by georglauterbach** [comment]
I will close this in favor of #2401.

**Comment by polarathene** [comment]
Just to add to the discussion. It was maybe related to syscall change with the newer Debian image, perhaps not due to Bash itself but another core package (**EDIT:** glibc AFAIK).

I don't recall the specifics but Alpine has a similar problem I became aware of a week or so ago (_**EDIT:** [this answer explains it](https://unix.stackexchange.com/questions/671246/why-is-bash-not-evaluating-the-executable-bit-correctly-in-alpine-3-14-2/671279#671279), for Debian 10 vs 11 it seems to be related to the glibc update, notably with kernels prior to 5.8_), where a change since their 3.14 release resulted in a syscall that older kernels didn't support properly.. I was surprised as it broke the `hash` command from bash from working correctly (_the Sentry.io CLI installer script was failing because of that_).

**Comment by georglauterbach** [comment]
Very interesting, so it was as I suspected earlier in the issue discussion (a kernel thing...). Thank you for sharing the information here, very much appreciated!

**Comment by oblitum** [comment]
@georglauterbach tbh I'm not sure how that relates to a leak due to subshell'ing, I couldn't connect the dots.

**Pull request closed without merging**